File: apps/backend/src/wallet/wallet.service.ts
import { Injectable, NotFoundException, UnprocessableEntityException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { WALLET_PACKAGES } from '@pawpals/shared';

@Injectable()
export class WalletService {
  constructor(private prisma: PrismaService) {}

  async getWallet(userId: string) {
    const wallet = await this.prisma.wallet.findUnique({ where: { userId } });
    if (!wallet) throw new NotFoundException('WALLET_NOT_FOUND');
    return { balance: wallet.balance, currency: '肉乾' };
  }

  async topup(userId: string, packageId: string) {
    const pkg = WALLET_PACKAGES.find((p) => p.id === packageId);
    if (!pkg) throw new UnprocessableEntityException('INVALID_PACKAGE');

    return this.prisma.$transaction(async (tx) => {
      const wallet = await tx.wallet.update({
        where: { userId },
        data: { balance: { increment: pkg.jerky } },
      });
      await tx.walletTransaction.create({
        data: { userId, type: 'TOPUP', amount: pkg.jerky, relatedEntityId: packageId },
      });
      return { balance: wallet.balance, currency: '肉乾' };
    });
  }

  async debitEscrow(userId: string, amount: number, meetingId: string) {
    return this.prisma.$transaction(async (tx) => {
      const wallet = await tx.wallet.findUnique({ where: { userId } });
      if (!wallet || wallet.balance < amount) throw new UnprocessableEntityException('INSUFFICIENT_BALANCE');
      await tx.wallet.update({ where: { userId }, data: { balance: { decrement: amount } } });
      await tx.walletTransaction.create({
        data: { userId, type: 'ESCROW', amount, relatedEntityId: meetingId },
      });
    });
  }

  async releaseEscrow(meetingId: string, recipientId: string, amount: number) {
    return this.prisma.$transaction(async (tx) => {
      await tx.wallet.update({ where: { userId: recipientId }, data: { balance: { increment: amount } } });
      await tx.walletTransaction.create({
        data: { userId: recipientId, type: 'ESCROW_RELEASE', amount, relatedEntityId: meetingId },
      });
    });
  }

  async refundEscrow(meetingId: string, userId: string, amount: number) {
    return this.prisma.$transaction(async (tx) => {
      await tx.wallet.update({ where: { userId }, data: { balance: { increment: amount } } });
      await tx.walletTransaction.create({
        data: { userId, type: 'CREDIT', amount, relatedEntityId: meetingId },
      });
    });
  }
}

File: apps/backend-java/src/main/java/com/pawpals/backend/wallet/WalletService.java
package com.pawpals.backend.wallet;

import com.pawpals.backend.common.ApiException;
import com.pawpals.backend.user.UserRepository;
import com.pawpals.backend.wallet.dto.WalletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class WalletService {

  private final WalletRepository walletRepository;
  private final WalletTransactionRepository transactionRepository;
  private final UserRepository userRepository;

  public WalletResponse getWallet(String userId) {
    Wallet wallet = walletRepository.findByUser_Id(userId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    return WalletResponse.of(wallet.getBalance());
  }

  @Transactional
  public WalletResponse topup(String userId, String packageId) {
    WalletPackages.Package pkg = WalletPackages.find(packageId);
    if (pkg == null) {
      throw ApiException.unprocessable("INVALID_PACKAGE");
    }

    Wallet wallet = walletRepository.lockByUserId(userId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    wallet.setBalance(wallet.getBalance() + pkg.jerky());
    walletRepository.save(wallet);

    recordTransaction(userId, WalletTransactionType.TOPUP, pkg.jerky(), packageId);
    return WalletResponse.of(wallet.getBalance());
  }

  @Transactional
  public void debitEscrow(String userId, int amount, String meetingId) {
    Wallet wallet = walletRepository.lockByUserId(userId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    if (wallet.getBalance() < amount) {
      throw ApiException.unprocessable("INSUFFICIENT_BALANCE");
    }
    wallet.setBalance(wallet.getBalance() - amount);
    walletRepository.save(wallet);
    recordTransaction(userId, WalletTransactionType.ESCROW, amount, meetingId);
  }

  @Transactional
  public void releaseEscrow(String meetingId, String recipientId, int amount) {
    Wallet wallet =
        walletRepository.lockByUserId(recipientId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    wallet.setBalance(wallet.getBalance() + amount);
    walletRepository.save(wallet);
    recordTransaction(recipientId, WalletTransactionType.ESCROW_RELEASE, amount, meetingId);
  }

  @Transactional
  public void refundEscrow(String meetingId, String userId, int amount) {
    Wallet wallet = walletRepository.lockByUserId(userId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    wallet.setBalance(wallet.getBalance() + amount);
    walletRepository.save(wallet);
    recordTransaction(userId, WalletTransactionType.CREDIT, amount, meetingId);
  }

  private void recordTransaction(String userId, WalletTransactionType type, int amount, String relatedEntityId) {
    WalletTransaction tx = new WalletTransaction();
    tx.setUser(userRepository.getReferenceById(userId));
    tx.setType(type);
    tx.setAmount(amount);
    tx.setRelatedEntityId(relatedEntityId);
    transactionRepository.save(tx);
  }
}
